source/credentials_provider_profile.c

```diff
@@ -45,52 +46,60 @@ static int s_profile_file_credentials_provider_get_credentials_async(
+                aws_string_c_str(impl->config_file_path));
+        } else {
+            AWS_LOGF_DEBUG(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "(id=%p) Profile credentials provider failed to build config profile collection from file at (%s)",
+                (void *)provider,
+                aws_string_c_str(impl->config_file_path));
+        }
 
-    /*
-     * Parse credentials file, if it exists
-     */
-    struct aws_profile_collection *credentials_profiles =
-        aws_profile_collection_new_from_file(provider->allocator, impl->credentials_file_path, AWS_PST_CREDENTIALS);
+        /*
+         * Parse credentials file, if it exists
+         */
+        struct aws_profile_collection *credentials_profiles =
+            aws_profile_collection_new_from_file(provider->allocator, impl->credentials_file_path, AWS_PST_CREDENTIALS);
 
-    if (credentials_profiles != NULL) {
-        AWS_LOGF_DEBUG(
-            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
-            "(id=%p) Profile credentials provider successfully built credentials profile collection from file at (%s)",
-            (void *)provider,
-            aws_string_c_str(impl->credentials_file_path));
-    } else {
-        AWS_LOGF_DEBUG(
-            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
-            "(id=%p) Profile credentials provider failed to build credentials profile collection from file at (%s)",
-            (void *)provider,
-            aws_string_c_str(impl->credentials_file_path));
+        if (credentials_profiles != NULL) {
+            AWS_LOGF_DEBUG(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "(id=%p) Profile credentials provider successfully built credentials profile collection from file at "
+                "(%s)",
+                (void *)provider,
+                aws_string_c_str(impl->credentials_file_path));
+        } else {
+            AWS_LOGF_DEBUG(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "(id=%p) Profile credentials provider failed to build credentials profile collection from file at (%s)",
+                (void *)provider,
+                aws_string_c_str(impl->credentials_file_path));
+        }
+
+        /*
+         * Merge the (up to) two sources into a single unified profile
+         */
+        merged_profiles =
+            aws_profile_collection_new_from_merge(provider->allocator, config_profiles, credentials_profiles);
+
+        aws_profile_collection_release(config_profiles);
+        aws_profile_collection_release(credentials_profiles);
     }
 
-    /*
-     * Merge the (up to) two sources into a single unified profile
-     */
-    struct aws_profile_collection *merged_profiles =
-        aws_profile_collection_new_from_merge(provider->allocator, config_profiles, credentials_profiles);
     if (merged_profiles != NULL) {
         const struct aws_profile *profile = aws_profile_collection_get_profile(merged_profiles, impl->profile_name);
```

Comment:
Maybe I didn't fully understand the profile collection, but, how can we tell the profile collection provided is a "merged" profile collection or not to verify that people has the right config passed in?

Reply:
It is up to the customers to pass a correct collection which contains everything we need. If it doesn't, we will throw a source failure error.

Reply:
> If it doesn't, we will throw a source failure error.

When we throw the failure? How will we throw it?

Reply:
We will throw an error if anything is missing from merged collection. Like [profile](https://github.com/awslabs/aws-c-auth/blob/8308060ab0d908a6242afdf369f6ea2d30a08e22/source/credentials_provider_profile.c#L95) is missing or credentials are missing https://github.com/awslabs/aws-c-auth/blob/8308060ab0d908a6242afdf369f6ea2d30a08e22/source/aws_profile.c#L34.